### PR TITLE
Add elastic apm blog

### PR DESCRIPTION
**Pull request by svenrienstra**

### CODE REVIEW

**Comment by bennycornelissen** [comment]
It seems the rendering of references/footnotes (e.g. `foo^1^`) doesn't look correct on the website. Apparently they work a little different in Hugo. Can you try the following? → https://www.markdownguide.org/extended-syntax/#footnotes

File: content/blog/elastic-apm-plugin.md [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+Today we'll be looking at an example were a Java application was using R2DBC (a database framework) and Elastic as the APM provider. The Elastic APM agent offers a plugin api so we can write our own instrumentation without needing to instrument each indiviual query. 
+
+### How does the APM agent work?
+
+Before we look into how we can write a plugin for our example case, let's dive into how the Elastic APM agent works. The Elastic APM agent uses Byte Buddy to instrument code, Byte Buddy can modify Java classes during runtime. This way the agent can change a class without recompiling it. What will typically happen is that the agent will add some code when a instrumented method is entered and exited.
```
Comment (bennycornelissen) [inline content/blog/elastic-apm-plugin.md]: As a full Java n00b I had never heard of Byte Buddy. Should we link to it?

File: content/blog/elastic-apm-plugin.md [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+Now that we know what to instrument, how do we actually record a query has been executed? Elastic is using OpenTelemetry^3^ to record events. In the terminology we call a request a 'trace' and within a trace we can have multiple or nested 'spans'. A span describes a single unit of work, for example a database query. OpenTelemtry also defines conventions of how to record data about the specific unit of work^4^.
```
Comment (bennycornelissen) [inline content/blog/elastic-apm-plugin.md]: Typo in 'OpenTelemetry'

**Review by bennycornelissen — CHANGES_REQUESTED** [review]
Small typo/formatting things, otherwise LGTM. Good post!

File: content/blog/elastic-apm-plugin.md [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+To record events on when for example a HTTP requests starts and ends, the agent will need to hook into the web framework of your choice. The agent provider generally will provide a list of supported technologies and frameworks. In the case of the HTTP request example the agent will dynamically add some code around the web framework to record when a HTTP request starts and ends.
+
+So what do you actually get by using these agents? In the example of Elastic APM you'll be able to see a timeline visualisation like below.
+
+{{<img src="https://www.elastic.co/guide/en/apm/guide/current/images/apm-distributed-tracing.png" class="img-fluid" title="Elastic APM dashboard" >}}
+
+We can see which HTTP request are fired and what database calls are being made. This information can be extremly useful when trying to understand what's happening within you application, for example when looking into a performance issue.
```
Comment (LFSb) [inline content/blog/elastic-apm-plugin.md]: Typo here in "within you application" 😄

File: content/blog/elastic-apm-plugin.md [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+  - Java
+  - Elastic APM
+  - R2DBC
+slug: elastic-apm-plugin
+date: 2023-07-12
+---
+
+As we all know observability is something we rely heavily on in our ever more complex landscape. In order to know if our applications are still doing what they should be doing we can't live without it. But also when something does go wrong it can provide us with very valuable insight of what is going on inside our application.
+
+One of the tools we can use to improve the observability of our applications are application performance monitoring agents (or APM agents). The agents provide information of what is going on inside our applications. They generally do this by instrumenting our code, without necessarily needing any changes within our own code. Most providers of APM tooling have agents available for a wide range of programming languages, such as Java, .Net, Python, PHP, etc. Using the instrumentation the agents can record events, like HTTP requests, database queries and messaging events.
```
Comment (LFSb) [inline content/blog/elastic-apm-plugin.md]: I'm being very very nitpicky here, but I personally like capitalizing the words that make up an initialism, so "Application Performance Monitoring (APM)" just reads nicer to me personally. Do with that what you will 😄

File: content/blog/elastic-apm-plugin.md [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+
+So what do you actually get by using these agents? In the example of Elastic APM you'll be able to see a timeline visualisation like below.
+
+{{<img src="https://www.elastic.co/guide/en/apm/guide/current/images/apm-distributed-tracing.png" class="img-fluid" title="Elastic APM dashboard" >}}
+
+We can see which HTTP request are fired and what database calls are being made. This information can be extremly useful when trying to understand what's happening within you application, for example when looking into a performance issue.
+
+### What if your framework is not supported?
+
+But what to do when your framework of choice is not supported by the agent. By not doing anything you might be missing out on valuable information, so is there anything we can do? Luckily there is! Most APM agents will offer some way of adding manual instrumentation in your code. This might be suitable if you want to add some instrumentation for a very specific use case. But what if for example your database framework is not supported? Adding manual instrumentation for each query is not an ideal solution.
```
Comment (LFSb) [inline content/blog/elastic-apm-plugin.md]: I'd say this first sentence should finish on a question mark.

File: content/blog/elastic-apm-plugin.md [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+
+Today we'll be looking at an example were a Java application was using R2DBC (a database framework) and Elastic as the APM provider. The Elastic APM agent offers a plugin api so we can write our own instrumentation without needing to instrument each indiviual query.
```
Comment (LFSb) [inline content/blog/elastic-apm-plugin.md]: Typo: "indiviual" -> "individual"

File: content/blog/elastic-apm-plugin.md (hunk 1 above)
Comment (LFSb) [reply content/blog/elastic-apm-plugin.md]: https://bytebuddy.net/#/ this one I guess? Never heard of it either, couldn't hurt to link it.

File: content/blog/elastic-apm-plugin.md [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+
+### Let's write our R2DBC plugin
+
+We now have a basic idea of how the APM agent works, so we can have a look at how we could write a plugin for our problem at hand. The aim is to be able to record queries have been executed, including SQL statement. We first need to identify what our entry point will be to instrument. The most obivous choice seems to be `io.r2dbc.spi.Statement#execute`^2^. According to the javadoc this method is responsible for `Executes one or more SQL statements and returns the Results.`. There is one problem however, the `Statement` API doesn't have any reference to the SQL statement. There are probably some ways around this, but to keep it simple we'll instrument a specific implementation: `io.r2dbc.postgresql.PostgresqlStatement`. This implementation has a `execute` method which takes the SQL statement as parameter.
```
Comment (LFSb) [inline content/blog/elastic-apm-plugin.md]: "The aim is to be able to record queries have been executed" -> "The aim is to be able to record queries **that** have been executed"

File: content/blog/elastic-apm-plugin.md [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+        return getTracer("r2dbc-postgresql")
+                .spanBuilder(signature.toString())
+                .setSpanKind(SpanKind.CLIENT)
+                .setAttribute("db.system", "postgresql")
+                .setAttribute("db.statement", sql);
+    }
+}
+```
+
+We wrap around the Reactor type here to start the span before we subscribe to the query result and once it completes we close the span. The signature parser used here is a copy of the `SignatureParser` class from the JDBC plugin of the APM agent^5^. To package the plugin a few more steps are needed, you can find them on the Elastic website: https://www.elastic.co/guide/en/apm/agent/java/current/plugin-api.html
```
Comment (LFSb) [inline content/blog/elastic-apm-plugin.md]: The end seems to be a bit abrupt, maybe a short summary of the above wouldn't be out of place here?

**Review by LFSb — COMMENTED** [review]
I found some typo's

**Comment by bennycornelissen** [comment]
Thanks @LFSb for _also_ proof reading. My brain is kinda fried so I missed like half the typos 😞

**Comment by LFSb** [comment]
> Thanks @LFSb for _also_ proof reading. My brain is kinda fried so I missed like half the typos 😞

I make the same mistakes on a daily basis so I'm extra alert for them 😛

File: content/blog/elastic-apm-plugin.md (hunk 1 above)
Comment (svenrienstra) [reply content/blog/elastic-apm-plugin.md]: Removed all reference to Byte Buddy as the exact isn't that important anyway

File: content/blog/elastic-apm-plugin.md (hunk 8 above)
Comment (svenrienstra) [reply content/blog/elastic-apm-plugin.md]: Yeah good point, the talk was starting so I rushed it a bit 😅

Comment (svenrienstra) [reply content/blog/elastic-apm-plugin.md]: Is this any better or did you expect something else ?

**Comment by svenrienstra** [comment]
Thanks for reviewing, I think I processed all comments :)

File: content/blog/elastic-apm-plugin.md [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+
+### Let's write our R2DBC plugin
+
+We now have a basic idea of how the APM agent works, so we can have a look at how we could write a plugin for our problem at hand. The aim is to be able to record queries that have been executed, including SQL statement. We first need to identify what our entry point will be to instrument. The most obvious choice seems to be `io.r2dbc.spi.Statement#execute[^2]. According to the javadoc this method is responsible for `Executes one or more SQL statements and returns the Results.`. There is one problem however, the `Statement` API doesn't have any reference to the SQL statement. There are probably some ways around this, but to keep it simple we'll instrument a specific implementation: `io.r2dbc.postgresql.PostgresqlStatement`. This implementation has a `execute` method which takes the SQL statement as parameter.
```
Comment (LFSb) [inline content/blog/elastic-apm-plugin.md]: I believe there's a backtick missing here, right before "According to the javadoc"

File: content/blog/elastic-apm-plugin.md [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+                .setAttribute("db.statement", sql);
+    }
+}
+```
+
+We wrap around the Reactor type here to start the span before we subscribe to the query result and once it completes we close the span. The span contains the statement (query) being executed. The query will now show up in our timeline (like in the screenshot earlier) and we'll be able to see a summary of the statement and how long it took to execute that statement.
+
+The signature parser that is being used is a copy of the `SignatureParser` class from the JDBC plugin of the APM agent[^5]. To package the plugin a few more steps are needed, you can find them on the Elastic website: https://www.elastic.co/guide/en/apm/agent/java/current/plugin-api.html
+
+
```
Comment (LFSb) [inline content/blog/elastic-apm-plugin.md:109]: Sadly, these links are not playing very nicely. Is there an alternative way of linking to these articles?

**Review by LFSb — COMMENTED** [review]
Nice! I have two last nitpicks, then I'm done 😄

**Comment by bennycornelissen** [comment]
I've fixed the footnotes 🎉 Looks good now in the Hugo preview in Gitpod, ~~so I expect the Netlify deploy preview to look the same when it's done building~~. Netlify deploy preview agrees: footnotes now work correctly 😊 

@svenrienstra @LFSb let's see if we can get this blog published today 🚀

**Review by bennycornelissen — APPROVED** [review]

File: content/blog/elastic-apm-plugin.md [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+To record events on when for example a HTTP requests starts and ends, the agent will need to hook into the web framework of your choice. The agent provider generally will provide a list of supported technologies and frameworks. In the case of the HTTP request example the agent will dynamically add some code around the web framework to record when a HTTP request starts and ends.
+
+So what do you actually get by using these agents? In the example of Elastic APM you'll be able to see a timeline visualisation like below.
+
+{{<img src="https://www.elastic.co/guide/en/apm/guide/current/images/apm-distributed-tracing.png" class="img-fluid" title="Elastic APM dashboard" >}}
```
Comment (bennycornelissen) [inline content/blog/elastic-apm-plugin.md]: Last nitpick: this is a potential 404 in the making. Shouldn't we mirror this image and have Netlify serve it for us? There's also some SEO thing (that I keep forgetting) with regard to using images in the blog metadata. Let me find that, so we can fix it in one go...

**Review by bennycornelissen — CHANGES_REQUESTED** [review]

File: content/blog/elastic-apm-plugin.md [hunk 12]
```diff
@@ -0,0 +1,113 @@
+---
+id: blog
+title: Writing a plugin for the Elastic APM Java agent
+description: This blog describes how to write a plugin for R2DBC for Elastic APM Java agent.
```
Comment (bennycornelissen) [inline content/blog/elastic-apm-plugin.md:4]: We need to add a `summary` block here as well. This is also used for the bit of text that's on the blog index. Without it, it just takes the first section of the blog and truncates that. Sometimes that works, sometimes.. not so much.

**Review by bennycornelissen — CHANGES_REQUESTED** [review]

File: content/blog/elastic-apm-plugin.md (hunk 12 above)
Comment (bennycornelissen) [reply content/blog/elastic-apm-plugin.md:4]: Proposal: 

```yaml
summary: Application Performance Monitoring (or APM) is one of today's powerful tools to get useful information about your application. Why is my application slow? What do I need to improve? There's a vast ecosystem of APM agents available with support for a wide range of languages and frameworks. But what do you do when your framework isn't supported? You create a plugin for your APM agent. In this blog we'll dive into doing just that.
```

File: content/blog/elastic-apm-plugin.md (hunk 11 above)
Comment (bennycornelissen) [reply content/blog/elastic-apm-plugin.md]: Looked into the SEO thing, where you also add an `images:` section to the metadata. It didn't look right so I propose we leave it out. I have a commit ready that brings the image local, so I'll just push that.

File: content/blog/elastic-apm-plugin.md (hunk 12 above)
Comment (bennycornelissen) [reply content/blog/elastic-apm-plugin.md:4]: @svenrienstra let me know if you're OK with this summary, or if you want to craft a better one 😊

**Review by LFSb — DISMISSED** [review]
LGTM!

File: content/blog/elastic-apm-plugin.md (hunk 12 above)
Comment (bennycornelissen) [reply content/blog/elastic-apm-plugin.md:4]: I've pushed my proposal. You're still free to ditch it if you want 😉

**Review by LFSb — APPROVED** [review]

**Comment by svenrienstra** [comment]
All changes look good to me, thanks for the help @bennycornelissen and @LFSb !

**Review by bennycornelissen — APPROVED** [review]
LGTM! Seems we all agree 😊 

image

**Pull request merged**